src/cswinrt/helpers.h

```diff
@@ -994,4 +994,202 @@ namespace cswinrt
+        if (exclusiveToAttribute)
+        {
+            auto sys_type = get_attribute_value<ElemSig::SystemType>(exclusiveToAttribute, 0);
+            TypeDef exclusiveToClass = iface.get_cache().find_required(sys_type.name);
+            if (!has_attribute(exclusiveToClass, "Windows.Foundation.Metadata"sv, "FastAbiAttribute"sv))
+            {
+                return {};
+            }
+            return exclusiveToClass;
+        }
+        return {};
+    }
+
+    std::pair<TypeDef, std::vector<TypeDef>> get_default_and_exclusive_interfaces(TypeDef const& classType)
+    {
+        std::pair<TypeDef, std::vector<TypeDef>> exclusive_ifaces;
+        std::vector<TypeDef> exclusive_interfaces;
+        for (auto&& ifaceImpl : classType.InterfaceImpl())
+        {
+            auto&& sem = get_type_semantics(ifaceImpl.Interface());
+            if (has_attribute(ifaceImpl, "Windows.Foundation.Metadata", "DefaultAttribute"))
+            {
+                exclusive_ifaces.first = std::get<type_definition>(sem);
+            }
+            else if (std::holds_alternative<type_definition>(sem))
+            {
+                if (has_attribute(std::get<type_definition>(sem), "Windows.Foundation.Metadata"sv, "ExclusiveToAttribute"sv))
+                {
+                    exclusive_ifaces.second.push_back(std::get<type_definition>(sem));
+                }
+            }
+        }
+        return exclusive_ifaces;
+    }
+
+    void sort_fast_abi_ifaces(std::vector<TypeDef>& fast_abi_ifaces)
+    {
+        std::sort(fast_abi_ifaces.begin(), fast_abi_ifaces.end(), [](TypeDef const& iface, TypeDef const& otherIface)
+        {
+            // compare relative contracts
+            auto relativeContractValueIface = -1 * get_number_of_attributes(iface, "Windows.Foundation.Metadata"sv, "PreviousContractVersionAttribute"sv);
+            auto relativeContractValueOtherIface = -1 * get_number_of_attributes(otherIface, "Windows.Foundation.Metadata"sv, "PreviousContractVersionAttribute"sv);
+            if (relativeContractValueIface != relativeContractValueOtherIface)
+                return relativeContractValueIface < relativeContractValueOtherIface;
+
+            //compare contract versions if they exist
+            auto contractVersionIface = get_contract_version(iface);
+            auto contractVersionOtherIface = get_contract_version(iface);
+            if (contractVersionIface.has_value() && contractVersionOtherIface.has_value() && contractVersionIface.value() != contractVersionOtherIface.value())
+                return contractVersionIface.value() < contractVersionOtherIface.value();
+
+            //compare versions
+            auto versionIface = get_version(iface);
+            auto versionOtherIface = get_version(iface);
+            if (versionIface.has_value() && versionOtherIface.has_value() && versionIface.value() != versionOtherIface.value())
```

Comment:
is this has value check guaranteed to return true for both.

Reply:
Did we confirm this?

Reply:
Yes, the winmd would either have version on both or none.